example/poisson/poisson_package.cpp

```diff
@@ -53,14 +54,140 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
   auto mphi = Metadata({Metadata::Cell, Metadata::Independent, Metadata::FillGhost});
   pkg->AddField("potential", mphi);
 
+  int ndim = 1 + (pin->GetInteger("parthenon/mesh", "nx2") > 1) +
+             (pin->GetInteger("parthenon/mesh", "nx3") > 1);
+  // set up the stencil object corresponding to the finite difference
+  // discretization we adopt in this pacakge
+  const int nstencil = 1 + 2 * ndim;
+  std::vector<std::vector<int>> offsets(
+      {{-1, 0, 1, 0, 0, 0, 0}, {0, 0, 0, -1, 1, 0, 0}, {0, 0, 0, 0, 0, -1, 1}});
+
+  bool use_jacobi = pin->GetOrAddBoolean("poisson", "use_jacobi", true);
+  pkg->AddParam<>("use_jacobi", use_jacobi);
+  bool use_stencil = pin->GetOrAddBoolean("poisson", "use_stencil", true);
+  pkg->AddParam<>("use_stencil", use_stencil);
+  if (use_stencil) {
+    std::vector<Real> wgts;
+    if (use_jacobi) {
+      wgts = std::vector<Real>({1.0, -2.0 * ndim, 1.0, 1.0, 1.0, 1.0, 1.0});
+    } else {
+      const Real w0 = 1.0 / (2.0 * ndim);
+      wgts = std::vector<Real>({w0, -1.0, w0, w0, w0, w0, w0});
+    }
+    auto stencil = parthenon::solvers::Stencil<Real>("stencil", nstencil, wgts, offsets);
+    pkg->AddParam<>("stencil", stencil);
+  } else {
+    // setup the sparse matrix
+    Metadata msp = Metadata({Metadata::Cell, Metadata::Derived, Metadata::OneCopy},
+                            std::vector<int>({nstencil}));
+    pkg->AddField("poisson_sparse_matrix", msp);
+    auto sp_accessor =
+        parthenon::solvers::SparseMatrixAccessor("accessor", nstencil, offsets);
+    pkg->AddParam("sparse_accessor", sp_accessor);
+  }
+
   return pkg;
 }
 
+template <typename T>
+TaskStatus SetMatrixElements(T *u) {
+  auto pm = u->GetParentPointer();
+
+  IndexRange ib = u->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = u->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = u->GetBoundsK(IndexDomain::interior);
+
+  PackIndexMap imap;
+  const std::vector<std::string> vars({"poisson_sparse_matrix"});
+  const auto &v = u->PackVariables(vars, imap);
+  const int isp_lo = imap["poisson_sparse_matrix"].first;
+  const int isp_hi = imap["poisson_sparse_matrix"].second;
+
+  if (isp_hi < 0) { // must be using the stencil so return
+    return TaskStatus::complete;
+  }
+
+  const int ndim = v.GetNdim();
+  const Real w0 = -2.0 * ndim;
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "SetMatElem", DevExecSpace(), 0, v.GetDim(5) - 1, kb.s, kb.e,
+      jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+        for (int n = isp_lo; n <= isp_hi; n++) {
+          v(b, n, k, j, i) = 1;
+        }
+        v(b, isp_lo + 1, k, j, i) = w0;
+      });
```

Comment:
I'm struggling to visualize the actual structure of the matrix v, what's it's full dimensions? `nblocks*27*nz*ny*nx`? So a few elements along the n dimension are 1 with the second being `w0`? And this matrix is repeated across all `k,j,i`?

Reply:
The point of these, at some level, is to not make you have to think about the actual sparse matrix.  They represent, very deliberately, the sparse matrices you get out of discretizing equations.  That's why you specify which elements are non-zero via offsets to the i,j,k index of a cell.  With that, it's super easy to fill one in for representing things like difference equations.  You'd never want to try to represent a generic sparse matrix with these.  The values above are what comes out of a simple centered finite difference discretization of the Laplacian.  Each `n` corresponds to either the cell itself or one of it's neighbors, as determined by the i,j,k offset arrays that are set up in the initialization routine.